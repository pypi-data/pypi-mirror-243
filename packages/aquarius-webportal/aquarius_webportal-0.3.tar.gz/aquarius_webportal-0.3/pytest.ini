[pytest]
log_cli = true
addopts = --doctest-modules --cov-config=.coveragerc --cov-report xml --cov-report term --cov=aquarius_webportal
testpaths = tests