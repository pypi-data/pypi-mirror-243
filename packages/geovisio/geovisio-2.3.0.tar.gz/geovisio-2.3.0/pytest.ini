[pytest]
markers =
	docker: mark a test as needing docker. Those tests will be slower than normal test
	skipci: mark a test a not running on ci. Is done for tests having complex dependencies, or very slow
