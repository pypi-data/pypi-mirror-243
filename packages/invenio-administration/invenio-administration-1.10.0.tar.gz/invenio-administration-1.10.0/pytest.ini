# -*- coding: utf-8 -*-
#
# Copyright (C) 2022 CERN.
#
# invenio-administration is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see LICENSE file for more
# details.

[pytest]
addopts = --isort --pydocstyle --pycodestyle --doctest-glob="*.rst" --doctest-modules --cov=invenio_administration --cov-report=term-missing
testpaths = docs tests invenio_administration
