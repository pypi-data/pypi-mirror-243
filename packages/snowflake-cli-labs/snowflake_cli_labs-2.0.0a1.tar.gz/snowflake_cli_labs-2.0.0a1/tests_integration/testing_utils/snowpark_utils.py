from __future__ import annotations

import json
import os
import re
from enum import Enum
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from snowflake.connector import SnowflakeConnection
from syrupy import SnapshotAssertion

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with
from tests_integration.testing_utils.assertions.test_file_assertions import (
    assert_that_current_working_directory_contains_only_following_files,
    assert_that_file_content_is_equal_to_snapshot,
)
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_contains_row_with,
    assert_that_result_is_successful,
    assert_that_result_is_successful_and_done_is_on_output,
    assert_that_result_is_successful_and_output_json_equals,
)
from tests_integration.testing_utils.file_utils import replace_text_in_file
from tests_integration.testing_utils.naming_utils import ObjectNameProvider
from tests_integration.testing_utils.sql_utils import SqlTestHelper

# TODO name 'entity' for both function and procedure does not seem to be perfect.
#  But i can`t come up with anything better


class TestType(Enum):
    FUNCTION = "function"
    PROCEDURE = "procedure"

    __test__ = False


class SnowparkTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
        sql_test_helper: SqlTestHelper,
        object_name_provider: ObjectNameProvider,
        test_database,
        test_type: TestType,
        snapshot: SnapshotAssertion,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session
        self.sql_test_helper = sql_test_helper
        self.test_object_name_provider = object_name_provider
        self.object_name_prefix = object_name_provider.get_object_name_prefix()
        self.test_type = test_type
        self.snapshot = snapshot

    def query_entities_created_in_this_test_case(self) -> List[Dict[str, Any]]:
        if self.test_type == TestType.FUNCTION:
            query_string = (
                f"SHOW USER FUNCTIONS LIKE '{self.object_name_prefix.upper()}%'"
            )
        else:
            query_string = f"SHOW PROCEDURES LIKE '{self.object_name_prefix.upper()}%'"

        return self.sql_test_helper.execute_single_sql(query_string)

    def query_stages_in_database(self) -> List[Dict[str, Any]]:
        return self.sql_test_helper.execute_single_sql(
            f"SHOW STAGES IN DATABASE {self.snowflake_session.database}"
        )

    def query_files_uploaded_in_this_test_case(
        self, stage_name: str
    ) -> List[Dict[str, Any]]:
        stages_in_database = [
            stage["name"].lower() for stage in self.query_stages_in_database()
        ]
        if stage_name in stages_in_database:
            return self.sql_test_helper.execute_single_sql(f"LIST @{stage_name}")
        else:
            return []

    def clean_after_test_case(self, stage_name: str) -> None:
        self.delete_entities_created_in_this_test_case()
        self.delete_files_uploaded_in_this_test_case(stage_name=stage_name)

    def delete_entities_created_in_this_test_case(self) -> None:
        entities_to_delete = self.query_entities_created_in_this_test_case()
        for obj in entities_to_delete:
            name_with_args = re.sub(r" RETURN .*", "", obj["arguments"])
            self.sql_test_helper.execute_single_sql(
                f"DROP {self.test_type.value.upper()} {name_with_args}"
            )

    def delete_files_uploaded_in_this_test_case(self, stage_name: str) -> None:
        files_to_remove = self.query_files_uploaded_in_this_test_case(
            stage_name=stage_name
        )
        for file_to_remove in files_to_remove:
            file_path_at_stage = file_to_remove["name"]
            self.sql_test_helper.execute_single_sql(f"REMOVE @{file_path_at_stage}")


class SnowparkTestSteps:
    dir_contents = {
        "function": [".gitignore", "app.py", "snowflake.yml", "requirements.txt"],
        "procedure": [
            "requirements.txt",
            "local_connection.py",
            ".gitignore",
            "app.py",
            "config.toml",
        ],
    }

    def __init__(self, setup: SnowparkTestSetup, test_type: TestType):
        self._setup = setup
        self.test_type = test_type

    def snowpark_list_should_return_no_data(self) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "list",
                "--like",
                f"{self._setup.object_name_prefix}%",
            ]
        )
        assert_that_result_is_successful_and_output_json_equals(result, [])

    def snowpark_list_should_return_entity_at_first_place(
        self, entity_name: str, arguments: str, result_type: str
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "list",
                "--like",
                f"{self._setup.object_name_prefix}%",
            ]
        )
        assert_that_result_is_successful(result)
        assert_that_result_contains_row_with(
            result,
            {
                "name": entity_name.upper(),
                "arguments": f"{entity_name}{arguments} RETURN {result_type}".upper(),
            },
        )

        if self.test_type == TestType.FUNCTION:
            assert_that_result_contains_row_with(result, {"language": "PYTHON"})

    def snowpark_execute_should_return_expected_value(
        self, entity_name: str, arguments: str, expected_value: Any
    ) -> None:
        identifier = entity_name + arguments
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "execute",
                identifier,
            ]
        )

        assert_that_result_is_successful(result)
        assert_that_result_contains_row_with(
            result,
            {
                identifier.upper()
                if self.test_type == TestType.FUNCTION
                else entity_name.upper(): expected_value
            },
        )

    def snowpark_describe_should_return_entity_description(
        self,
        entity_name: str,
        arguments: str,
        signature: Optional[str] = None,
        returns: Optional[str] = None,
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "describe",
                entity_name + arguments,
            ]
        )
        assert_that_result_is_successful(result)

        if returns:
            assert_that_result_contains_row_with(
                result, {"property": "returns", "value": returns}
            )
        assert_that_result_contains_row_with(
            result, {"property": "signature", "value": signature or arguments}
        )
        assert result.json is not None

    def snowpark_init_should_initialize_files_with_default_content(
        self,
    ) -> None:
        result = self._setup.runner.invoke_json(
            ["snowpark", self.test_type.value, "init", ".", "--format", "JSON"]
        )
        file_list = self.dir_contents[self.test_type.value]

        assert result.output is not None
        assert json.loads(result.output) == {
            "message": "Initialized the new project in ./"
        }
        assert set(os.listdir()) == set(file_list)

        for file in file_list:
            assert_that_file_content_is_equal_to_snapshot(
                actual_file_path=file, snapshot=self._setup.snapshot(name=file)
            )

    def add_parameters_to_procedure(self, parameters: str):
        replace_text_in_file(
            file_path="app.py",
            to_replace="def hello(session: Session) -> str:",
            replacement=f"def hello(session: Session, {parameters}) -> str:",
        )

    def requirements_file_should_contain_coverage(self, file_name="requirements.txt"):
        assert os.path.exists(file_name)

        with open(file_name, "r") as req_file:
            file_contents = req_file.readlines()
            assert file_contents
            assert "coverage\n" in file_contents

    def snowpark_package_should_zip_files(self) -> None:
        value_to_cmd = {"function": "build", "procedure": "package"}
        result = self._setup.runner.invoke_json(
            [
                "snowpark",
                self.test_type.value,
                value_to_cmd[self.test_type.value],
                "--pypi-download",
                "yes",
                "--format",
                "JSON",
            ]
        )
        assert_that_result_is_successful_and_done_is_on_output(result)
        assert_that_current_working_directory_contains_only_following_files(
            *self.dir_contents[self.test_type.value],
            "app.zip",
            "requirements.snowflake.txt",
        )

    def snowpark_create_should_finish_successfully(self, entity_name: str) -> str:
        return self.run_deploy(entity_name)

    def snowpark_deploy_with_coverage_wrapper_should_finish_successfully(
        self, entity_name: str
    ) -> str:
        return self.run_deploy(entity_name, "--install-coverage-wrapper")

    def run_deploy(self, entity_name: str, additional_arguments: str = ""):

        arguments = [
            "snowpark",
            self.test_type.value,
            "deploy",
            entity_name,
        ]

        if additional_arguments:
            arguments.append(additional_arguments)

        result = self._setup.runner.invoke_with_connection_json(arguments)

        assert_that_result_is_successful(result)
        return entity_name

    def snowpark_deploy_should_not_replace_if_the_signature_does_not_change(
        self, entity_name: str
    ):
        replace_text_in_file(
            file_path="app.py",
            to_replace='return f"Hello {name}!"',
            replacement='return "Hello Snowflakes!"',
        )

        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "deploy",
                entity_name,
                "--replace",
            ]
        )

        assert_that_result_is_successful_and_output_json_equals(
            result, [{"message": "No packages to update. Deployment complete!"}]
        )

    def snowpark_deploy_should_finish_successfully(
        self,
        entity_name: str,
    ) -> None:
        replace_text_in_file(
            file_path="app.py",
            to_replace="def hello() -> str:",
            replacement="def hello() -> int:",
        )
        replace_text_in_file(
            file_path="app.py",
            to_replace='return "Hello Snowflakes!"',
            replacement="return 1",
        )
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "deploy",
                entity_name,
                "--replace",
            ]
        )
        assert_that_result_is_successful_and_output_json_equals(
            result,
            {"status": f"Function {entity_name.upper()} successfully created."},
        )

    def snowpark_drop_should_finish_successfully(
        self,
        entity_name: str,
        arguments: str,
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                self.test_type.value,
                "drop",
                entity_name + arguments,
            ]
        )
        assert_that_result_is_successful(result)

    def procedure_coverage_should_return_report_when_files_are_present_on_stage(
        self, identifier: str
    ):
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "procedure",
                "coverage",
                "report",
                identifier,
                "--output-format",
                "json",
            ]
        )

        assert result.exit_code == 0
        assert os.path.isfile("coverage.json")

        with open("coverage.json", "r") as coverage_file:
            coverage = json.load(coverage_file)

        assert "percent_covered" in coverage["totals"].keys()
        assert "excluded_lines" in coverage["totals"].keys()

    def coverage_clear_should_execute_successfully(self, identifier: str):
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "procedure",
                "coverage",
                "clear",
                identifier,
            ]
        )

        assert result.exit_code == 0
        assert result.json["result"] == "removed"  # type: ignore

    def assert_that_no_entities_are_in_snowflake(self) -> None:
        self.assert_that_only_these_entities_are_in_snowflake()

    def assert_that_only_these_entities_are_in_snowflake(
        self, *expected_full_entity_signatures: str
    ) -> None:
        actual_entity_signatures = [
            entity["arguments"]
            for entity in self._setup.query_entities_created_in_this_test_case()
        ]
        adjusted_expected_full_function_signatures = [
            entity.upper() for entity in expected_full_entity_signatures
        ]
        assert set(actual_entity_signatures) == set(
            adjusted_expected_full_function_signatures
        )

    def assert_that_no_files_are_staged_in_test_db(self) -> None:
        self.assert_that_only_these_files_are_staged_in_test_db()

    def assert_that_only_these_files_are_staged_in_test_db(
        self, *expected_file_paths: str
    ) -> None:
        f = self.get_actual_files_staged_in_db()
        assert set(self.get_actual_files_staged_in_db()) == set(expected_file_paths)

    def assert_that_only_app_and_coverage_file_are_staged_in_test_db(
        self, path_beggining: str
    ):
        coverage_regex = re.compile(
            path_beggining + "/coverage/[0-9]{8}-[0-9]{6}.coverage"
        )
        app_name = path_beggining + "/app.zip"

        assert app_name in (actual_file_paths := self.get_actual_files_staged_in_db())
        assert any(coverage_regex.match(file) for file in actual_file_paths)

    def get_actual_files_staged_in_db(self):
        return [
            staged_file["name"]
            for staged_file in self._setup.query_files_uploaded_in_this_test_case()
        ]

    @staticmethod
    def add_requirements_to_requirements_txt(
        requirements: List[str], file_path: str = "requirements.txt"
    ):
        if os.path.exists(file_path):
            with open(file_path, "a") as reqs_file:
                for req in requirements:
                    reqs_file.write(req + "\n")

    def get_entity_name(self):
        return self._setup.test_object_name_provider.create_and_get_next_object_name()


# Temporary copy for procedures until will be switched to project definition
class SnowparkProcedureTestSteps:
    def __init__(self, setup: SnowparkTestSetup):
        self._setup = setup
        self.file_dir_list = {
            Path("app"),
            Path("requirements.txt"),
            Path(".gitignore"),
            Path("snowflake.yml"),
            Path("app/functions.py"),
            Path("app/__init__.py"),
            Path("app/common.py"),
            Path("app/procedures.py"),
        }

    def object_show_should_return_no_data(self, object_type: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                object_type,
                "--like",
                f"{self._setup.object_name_prefix}%",
            ]
        )
        assert_that_result_is_successful_and_output_json_equals(result, [])

    def object_show_includes_given_identifiers(
        self,
        object_type: str,
        identifier: Tuple[str, str],
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                object_type,
                "--like",
                f"{self._setup.object_name_prefix}%",
            ]
        )
        assert_that_result_is_successful(result)

        entity_name, signature = identifier
        assert_that_result_contains_row_with(
            result,
            {
                "name": entity_name.upper(),
                "arguments": f"{entity_name}{signature}".upper(),
            },
        )

        if object_type == TestType.FUNCTION:
            assert_that_result_contains_row_with(result, {"language": "PYTHON"})

    def snowpark_execute_should_return_expected_value(
        self, object_type: str, identifier: str, expected_value: Any
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "execute",
                object_type,
                identifier,
            ]
        )

        entity_name = identifier.rsplit("(")[0]

        assert_that_result_is_successful(result)
        assert_that_result_contains_row_with(
            result,
            {
                identifier.upper()
                if object_type == TestType.FUNCTION.value
                else entity_name.upper(): expected_value
            },
        )

    def object_describe_should_return_entity_description(
        self,
        object_type: str,
        identifier: str,
        signature: str,
        returns: str,
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "describe",
                object_type,
                identifier,
            ]
        )
        assert_that_result_is_successful(result)

        if returns:
            assert_that_result_contains_row_with(
                result, {"property": "returns", "value": returns}
            )
        assert_that_result_contains_row_with(
            result, {"property": "signature", "value": signature}
        )
        assert result.json is not None

    def snowpark_init_should_initialize_files_with_default_content(
        self,
    ) -> None:
        result = self._setup.runner.invoke_json(
            ["snowpark", "init", ".", "--format", "JSON"]
        )
        assert_that_result_is_successful(result)
        assert result.json == {"message": "Initialized the new project in ./"}

        current_files = set(Path(".").glob("**/*"))
        assert current_files == self.file_dir_list

        for file in current_files:
            if not file.is_file():
                continue
            assert_that_file_content_is_equal_to_snapshot(
                actual_file_path=file, snapshot=self._setup.snapshot(name=file.name)
            )

    def add_parameters_to_procedure(self, parameters: str):
        replace_text_in_file(
            file_path="app.py",
            to_replace="def hello(session: Session, name: str) -> str:",
            replacement=f"def hello(session: Session, {parameters}) -> str:",
        )

    def requirements_file_should_contain_coverage(self, file_name="requirements.txt"):
        assert os.path.exists(file_name)

        with open(file_name, "r") as req_file:
            file_contents = req_file.readlines()
            assert file_contents
            assert "coverage\n" in file_contents

    def snowpark_package_should_zip_files(self) -> None:
        current_files = set(Path(".").glob("**/*"))
        result = self._setup.runner.invoke_json(
            [
                "snowpark",
                "build",
                "--pypi-download",
                "yes",
                "--format",
                "JSON",
            ]
        )
        assert result.exit_code == 0, result.output
        assert result.json, result.output
        assert "message" in result.json
        assert "Build done. Artefact path:" in result.json["message"]  # type: ignore

        assert_that_current_working_directory_contains_only_following_files(
            *current_files,
            Path("app.zip"),
            Path("requirements.snowflake.txt"),
        )

    def snowpark_deploy_with_coverage_wrapper_should_finish_successfully(
        self,
    ):
        return self.run_deploy("--install-coverage-wrapper")

    def run_deploy(self, additional_arguments: str = ""):
        arguments = [
            "snowpark",
            "deploy",
        ]

        if additional_arguments:
            arguments.append(additional_arguments)

        result = self._setup.runner.invoke_with_connection_json(arguments)
        return result

    def snowpark_deploy_should_not_replace_if_the_signature_does_not_change(
        self, entity_name: str
    ):
        replace_text_in_file(
            file_path="app.py",
            to_replace='return "Hello World!"',
            replacement='return "Hello Snowflakes!"',
        )

        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "deploy",
            ]
        )

        assert_that_result_is_successful_and_output_json_equals(
            result, {"message": "No packages to update. Deployment complete!"}
        )

    def snowpark_deploy_should_finish_successfully(
        self,
        entity_name: str,
    ) -> None:
        replace_text_in_file(
            file_path="app.py",
            to_replace="def hello() -> str:",
            replacement="def hello() -> int:",
        )
        replace_text_in_file(
            file_path="app.py",
            to_replace='return "Hello Snowflakes!"',
            replacement="return 1",
        )
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "deploy",
                entity_name + "()",
                "--handler",
                "app.hello",
                "--returns",
                "int",
                "--replace",
            ]
        )
        assert_that_result_is_successful_and_output_json_equals(
            result,
            {"status": f"Function {entity_name.upper()} successfully created."},
        )

    def object_drop_should_finish_successfully(
        self, object_type: str, identifier: str
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                object_type,
                identifier,
            ]
        )
        assert_that_result_is_successful(result)

    def procedure_coverage_should_return_report_when_files_are_present_on_stage(
        self, identifier: str
    ):
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "coverage",
                "report",
                identifier,
                "--output-format",
                "json",
            ]
        )

        assert result.exit_code == 0
        assert os.path.isfile("coverage.json")

        with open("coverage.json", "r") as coverage_file:
            coverage = json.load(coverage_file)

        assert "percent_covered" in coverage["totals"].keys()
        assert "excluded_lines" in coverage["totals"].keys()

    def coverage_clear_should_execute_successfully(self):
        result = self._setup.runner.invoke_with_connection_json(
            [
                "snowpark",
                "coverage",
                "clear",
            ]
        )

        assert result.exit_code == 0
        assert result.json["result"] == "removed"  # type: ignore

    def assert_those_procedures_are_in_snowflake(
        self, *expected_full_entity_signatures: str
    ):
        query_string = (
            f"SHOW PROCEDURES LIKE '{self._setup.object_name_prefix.upper()}%'"
        )
        result = self._setup.sql_test_helper.execute_single_sql(query_string)
        self.assert_that_only_these_entities_are_in_snowflake(
            result, *expected_full_entity_signatures
        )

    def assert_those_functions_are_in_snowflake(
        self, *expected_full_entity_signatures: str
    ):
        query_string = (
            f"SHOW USER FUNCTIONS LIKE '{self._setup.object_name_prefix.upper()}%'"
        )
        result = self._setup.sql_test_helper.execute_single_sql(query_string)
        self.assert_that_only_these_entities_are_in_snowflake(
            result, *expected_full_entity_signatures
        )

    def assert_no_procedures_in_snowflake(self):
        self.assert_those_procedures_are_in_snowflake()

    def assert_no_functions_in_snowflake(self):
        self.assert_those_functions_are_in_snowflake()

    def assert_that_only_these_entities_are_in_snowflake(
        self, query_result, *expected_full_entity_signatures: str
    ) -> None:
        actual_entity_signatures = [entity["arguments"] for entity in query_result]
        adjusted_expected_full_function_signatures = [
            entity.upper() for entity in expected_full_entity_signatures
        ]
        assert set(actual_entity_signatures) == set(
            adjusted_expected_full_function_signatures
        )

    def assert_that_no_files_on_stage(self, stage_name: str) -> None:
        self.assert_that_only_these_files_are_staged_in_test_db(stage_name=stage_name)

    def assert_that_only_these_files_are_staged_in_test_db(
        self,
        *expected_file_paths: str,
        stage_name: str,
    ) -> None:
        assert set(self.get_actual_files_staged_in_db(stage_name)) == set(
            expected_file_paths
        )

    def assert_that_only_app_and_coverage_file_are_staged_in_test_db(
        self, stage_path: str, artifact_name: str, stage_name: str
    ):
        coverage_regex = re.compile(stage_path + "/coverage/[0-9]{8}-[0-9]{6}.coverage")

        assert f"{stage_path}/{artifact_name}" in (
            actual_file_paths := self.get_actual_files_staged_in_db(
                stage_name=stage_name
            )
        )
        assert any(coverage_regex.match(file) for file in actual_file_paths)

    def get_actual_files_staged_in_db(self, stage_name: str):
        return [
            staged_file["name"]
            for staged_file in self._setup.query_files_uploaded_in_this_test_case(
                stage_name=stage_name
            )
        ]

    @staticmethod
    def add_requirements_to_requirements_txt(
        requirements: List[str], file_path: str = "requirements.txt"
    ):
        if os.path.exists(file_path):
            with open(file_path, "a") as reqs_file:
                for req in requirements:
                    reqs_file.write(req + "\n")

    def get_entity_name(self):
        return self._setup.test_object_name_provider.create_and_get_next_object_name()
