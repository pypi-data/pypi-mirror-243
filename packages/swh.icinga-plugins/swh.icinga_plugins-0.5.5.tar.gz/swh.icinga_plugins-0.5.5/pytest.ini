[pytest]
addopts = -p no:django
norecursedirs = build docs .*

asyncio_mode = strict
