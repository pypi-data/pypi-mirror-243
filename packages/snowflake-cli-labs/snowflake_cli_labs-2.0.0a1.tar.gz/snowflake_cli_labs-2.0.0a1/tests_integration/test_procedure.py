from __future__ import annotations

from pathlib import Path
from time import sleep

import pytest

from tests_integration.testing_utils import assert_that_result_is_successful
from tests_integration.testing_utils.snowpark_utils import (
    SnowparkProcedureTestSteps,
    SnowparkTestSetup,
    TestType,
)

STAGE_NAME = "dev_deployment"


@pytest.mark.integration
def test_snowpark_procedure_flow(_test_steps, project_directory, alter_snowflake_yml):
    _test_steps.assert_no_procedures_in_snowflake()
    _test_steps.assert_no_functions_in_snowflake()

    _test_steps.assert_that_no_files_on_stage(stage_name=STAGE_NAME)

    _test_steps.object_show_should_return_no_data(object_type="function")
    _test_steps.object_show_should_return_no_data(object_type="procedure")

    procedure_name = _test_steps.get_entity_name()
    function_name = _test_steps.get_entity_name()

    with project_directory("snowpark") as tmp_dir:
        project_file: Path = tmp_dir / "snowflake.yml"
        _test_steps.snowpark_package_should_zip_files()

        alter_snowflake_yml(
            project_file,
            parameter_path="snowpark.procedures.0.name",
            value=procedure_name,
        )
        alter_snowflake_yml(
            project_file,
            parameter_path="snowpark.functions.0.name",
            value=function_name,
        )

        result = _test_steps.run_deploy()
        assert_that_result_is_successful(result)
        assert result.json == [
            {
                "object": f"{procedure_name}(name string)",
                "status": "created",
                "type": "procedure",
            },
            {"object": "test()", "status": "created", "type": "procedure"},
            {
                "object": f"{function_name}(name string)",
                "status": "created",
                "type": "function",
            },
        ]

        _test_steps.assert_those_procedures_are_in_snowflake(
            f"{procedure_name}(VARCHAR) RETURN VARCHAR"
        )
        _test_steps.assert_those_functions_are_in_snowflake(
            f"{function_name}(VARCHAR) RETURN VARCHAR"
        )

        expected_files = [
            f"{STAGE_NAME}/my_snowpark_project/app.zip",
        ]
        _test_steps.assert_that_only_these_files_are_staged_in_test_db(
            *expected_files, stage_name=STAGE_NAME
        )

        # Listing procedures or functions shows created objects
        _test_steps.object_show_includes_given_identifiers(
            object_type="procedure",
            identifier=(procedure_name, "(VARCHAR) RETURN VARCHAR"),
        )
        _test_steps.object_show_includes_given_identifiers(
            object_type="function",
            identifier=(function_name, "(VARCHAR) RETURN VARCHAR"),
        )

        # Created objects can be described
        _test_steps.object_describe_should_return_entity_description(
            object_type="procedure",
            identifier=f"{procedure_name}(VARCHAR)",
            signature="(NAME VARCHAR)",
            returns="VARCHAR(16777216)",
        )

        _test_steps.object_describe_should_return_entity_description(
            object_type="function",
            identifier=f"{function_name}(VARCHAR)",
            signature="(NAME VARCHAR)",
            returns="VARCHAR(16777216)",
        )

        # Created objects can be executed
        _test_steps.snowpark_execute_should_return_expected_value(
            object_type="procedure",
            identifier=f"{procedure_name}('foo')",
            expected_value="Hello foo",
        )

        _test_steps.snowpark_execute_should_return_expected_value(
            object_type="function",
            identifier=f"{function_name}('foo')",
            expected_value="Hello foo!",
        )

        # Subsequent deploy of same object should fail
        result = _test_steps.run_deploy()
        assert result.exit_code == 1
        assert "already exists" in result.output

        # Apply changes to project objects
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path="snowpark.procedures.0.returns",
            value="variant",
        )
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path="snowpark.functions.0.returns",
            value="variant",
        )

        # Now we deploy with replace flag, it should update existing objects
        result = _test_steps.run_deploy("--replace")
        assert_that_result_is_successful(result)
        assert result.json == [
            {
                "object": f"{procedure_name}(name string)",
                "status": "definition updated",
                "type": "procedure",
            },
            {
                "object": "test()",
                "status": "packages updated",
                "type": "procedure",
            },
            {
                "object": f"{function_name}(name string)",
                "status": "definition updated",
                "type": "function",
            },
        ]

        # Check if objects were updated
        _test_steps.assert_those_procedures_are_in_snowflake(
            f"{procedure_name}(VARCHAR) RETURN VARIANT"
        )
        _test_steps.assert_those_functions_are_in_snowflake(
            f"{function_name}(VARCHAR) RETURN VARIANT"
        )

        # Same file should be present
        _test_steps.assert_that_only_these_files_are_staged_in_test_db(
            *expected_files, stage_name=STAGE_NAME
        )

        # Listing procedures or functions shows updated objects
        _test_steps.object_show_includes_given_identifiers(
            object_type="procedure",
            identifier=(procedure_name, "(VARCHAR) RETURN VARIANT"),
        )
        _test_steps.object_show_includes_given_identifiers(
            object_type="function",
            identifier=(function_name, "(VARCHAR) RETURN VARIANT"),
        )

        # Updated objects can be executed
        _test_steps.snowpark_execute_should_return_expected_value(
            object_type="procedure",
            identifier=f"{procedure_name}('foo')",
            expected_value='"Hello foo"',
        )

        _test_steps.snowpark_execute_should_return_expected_value(
            object_type="function",
            identifier=f"{function_name}('foo')",
            expected_value='"Hello foo!"',
        )

        # Check if objects can be dropped
        _test_steps.object_drop_should_finish_successfully(
            object_type="procedure", identifier=f"{procedure_name}(varchar)"
        )
        _test_steps.object_drop_should_finish_successfully(
            object_type="function", identifier=f"{function_name}(varchar)"
        )

        _test_steps.object_show_should_return_no_data(object_type="function")
        _test_steps.object_show_should_return_no_data(object_type="procedure")

        _test_steps.assert_that_only_these_files_are_staged_in_test_db(
            *expected_files, stage_name=STAGE_NAME
        )


@pytest.fixture
def _test_setup(
    runner,
    snowflake_session,
    sql_test_helper,
    object_name_provider,
    test_database,
    temporary_working_directory,
    snapshot,
):
    snowpark_procedure_test_setup = SnowparkTestSetup(
        runner=runner,
        snowflake_session=snowflake_session,
        sql_test_helper=sql_test_helper,
        object_name_provider=object_name_provider,
        test_database=test_database,
        snapshot=snapshot,
        test_type=TestType.PROCEDURE,
    )
    yield snowpark_procedure_test_setup
    snowpark_procedure_test_setup.clean_after_test_case(stage_name=STAGE_NAME)


@pytest.fixture
def _test_steps(_test_setup):
    yield SnowparkProcedureTestSteps(_test_setup)
