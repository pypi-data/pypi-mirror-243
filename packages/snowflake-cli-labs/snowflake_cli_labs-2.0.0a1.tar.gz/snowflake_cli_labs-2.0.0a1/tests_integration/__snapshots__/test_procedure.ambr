# serializer version: 1
# name: test_snowpark_procedure_flow[.gitignore]
  '''
  .packages/
  .venv/
  app.zip
  __pycache__
  
  '''
# ---
# name: test_snowpark_procedure_flow[__init__.py]
  ''
# ---
# name: test_snowpark_procedure_flow[common.py]
  '''
  def print_hello(name: str):
      return f"Hello {name}!"

  '''
# ---
# name: test_snowpark_procedure_flow[functions.py]
  '''
  from __future__ import annotations

  from common import print_hello


  def hello_function(name: str) -> str:
      return print_hello(name)

  '''
# ---
# name: test_snowpark_procedure_flow[procedures.py]
  '''
  from __future__ import annotations
  
  from common import print_hello
  from snowflake.snowpark import Session
  
  
  def hello_procedure(session: Session, name: str) -> str:
      token = _snowflake.get_oauth_access_token("cred")
      return print_hello(name)
  
  
  def test_procedure(session: Session) -> str:
      token = _snowflake.get_oauth_access_token("cred")
      return "Test procedure"

  '''
# ---
# name: test_snowpark_procedure_flow[requirements.txt]
  '''
  snowflake-snowpark-python
  
  '''
# ---
# name: test_snowpark_procedure_flow[snowflake.yml]
  '''
  definition_version: 1
  snowpark:
    project_name: "my_snowpark_project"
    stage_name: "dev_deployment"
    src: "app/"
    functions:
      - name: hello_function
        handler: "functions.hello_function"
        signature:
          - name: "name"
            type: "string"
        returns: string
    procedures:
      - name: hello_procedure
        handler: "procedures.hello_procedure"
        signature:
          - name: "name"
            type: "string"
        returns: string
        external_access_integration:
          - my_external_access
        secrets:
          cred: my_cred_name
      - name: test_procedure
        handler: "procedures.test_procedure"
        signature: ""
        returns: string
  
  '''
# ---
